### DISCUSSION TITLE
RTU client => ESP32 BRIDGE => TCP server

### DISCUSSION THREAD

**Discussion by Oo87** [eModbus/eModbus#211] [Q&A] [answered]
Hello all,

I have now 2 ESP32 working as Bridges from RTU server => ESP32 bridge => TCP client. working just great.

I'm now stucked doing the same the other way : RTU Client => ESP32 bridge => TCP server

Basic aplication is my solar inverter (RTU client) is far away from the energy meter(rtu server, also TCP server thanks to the bridge).

```
// Set up ModbusTCP client.
// - provide onData handler function
  MB.onDataHandler(&handleData);
// - provide onError handler function
  MB.onErrorHandler(&handleError);
// Set message timeout to 2000ms and interval between requests to the same host to 200ms
  MB.setTimeout(2000, 200);
// Start ModbusTCP background task
  MB.begin();

  MB.setTarget(IPAddress(192, 168, 5, 212), 502);

  Error err = MB.addRequest((uint32_t)millis(), 1, READ_HOLD_REGISTER, 10, 4);
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }
```
this works perfect, got data.

```
// BRIDGE SETUP
// bool attachServer(uint8_t aliasID, uint8_t serverID, uint8_t functionCode, ModbusClient *client, IPAddress host, uint16_t port)
// ServerID 5: Server with remote serverID 4, accessed through RTU client MB
//             Only FCs 03 and 04 accepted
  MBbridge.attachServer(1, 1, ANY_FUNCTION_CODE , &MB,IPAddress(192, 168, 5, 212),502);
  MBbridge.addFunctionCode(5, READ_INPUT_REGISTER);

// Check: print out all combinations served to Serial
  MBbridge.listServer();

// Start the bridge. Port 502, 4 simultaneous clients allowed, 600ms inactivity to disconnect client
  MBbridge.start(port, 4, 600);
```

this doesn't make anything

Anybody can help me how to setup this ?

Thank you for helping :)

**Comment by Miq1** [comment]
It looks like a server ID mismatch to me. You are attaching the TCP server with ID 1 to the bridge as server with again the ID 1. So far, so plausible. But then in the next line you add the READ_INPUT_REGISTER function code for the server ID 5, that is not known to the bridge yet.
Could it be you rather meant ``MBridge.addFunctionCode(1, READ_INPUT_REGISTER);`` instead? (that should be not necessary, though, as you had ANY_FUNCTION_CODE in the attach)

**Reply by Oo87** [reply]
Thanks for your help !

indeed, this is still an wrong copy paste line from the example :s .

since all my client and server are using adress 1 and they are always alone on their bus. what should I write ?

`MBbridge.attachServer(1, 1, ANY_FUNCTION_CODE , &MB,IPAddress(192, 168, 5, 212),502);`

also,

My code have those lines too : 

```
// SERVER RTU SETUP
// Init Serial2 conneted to the RTU Modbus
// (Fill in your data here!)
  Serial2.begin(9600, SERIAL_8N1, GPIO_NUM_16, GPIO_NUM_17);
// Register served function code worker for server 1, FC 0x03
  MBserver.registerWorker(0x01, ANY_FUNCTION_CODE, &FC03);
  MBserver.registerWorker(0x08, READ_HOLD_REGISTER, &FC03);

// Start ModbusRTU background task
  MBserver.start();
```

Do we have too setup also the MBserver? or the bridge handle it?

**Reply by Miq1** [reply]
The ``attachServer()`` call does make the TCP server with ID 1 accessible with ID 1 as well, if a client is requesting from the bridge.
I suppose the ``MBserver`` is defined for another ESP32 rather than for the bridge, where the bridge is called ``MBbridge``? It would make no sense to have more than one server on a device normally, and the ``MBbridge`` object has all features and methods of a Modbus server itself.
Or do you attempt to run the same ESP32 both as a TCP and a RTU bridge simultaneously?

**Reply by Oo87** [reply]
Or do you attempt to run the same ESP32 both as a TCP and a RTU bridge simultaneously?
=> no.

i use MBServer to see that wiring,etc works.
```
[D] Raw buffer received: @3FFCE504/8:
  | 0000: 02 04 00 12 00 12 D0 31                           |.......1        |
[D] Received packet: @3FFCCBF4/6:
  | 0000: 02 04 00 12 00 12                                 |......          |
[D] 144453| ModbusServerRTU.cpp  [ 155] serve: Request received.
[D] 144459| ModbusServer.cpp     [  44] getWorker: No matching worker found
[V] 145038| RTUutils.cpp         [ 289] receive: 3672us without data
```
And it works, of course no worker configured here but if I do MBserver deal with the request from the RTU client.

also this : 

```
// Set up ModbusTCP client.
// - provide onData handler function
  MB.onDataHandler(&handleData);
// - provide onError handler function
  MB.onErrorHandler(&handleError);
// Set message timeout to 2000ms and interval between requests to the same host to 200ms
  MB.setTimeout(2000, 200);
// Start ModbusTCP background task
  MB.begin();

  MB.setTarget(IPAddress(192, 168, 5, 212), 502);

  Error err = MB.addRequest((uint32_t)millis(), 1, READ_HOLD_REGISTER, 10, 4);
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }
```
It also works, I get data from TCP server (so requested as tcp client.

Now I just wonder how to link this RTUserver to this TCP client.

Network is : 

```
##### ESP32 #1 #### >>>>>>>>>>>> ###### ESP32 #2 #########
TCP server adress 1 <=> WIFI <=> TCPclient+Bridge+RTUserver <=> RTU (inverter)
```

**Reply by Oo87** [reply]
I also dont understand this line : 

// Start the bridge. Port 502, 4 simultaneous clients allowed, 600ms inactivity to disconnect client
  MBbridge.start(port, 4, 600);
Since connection come from RTU and not TCP. because TCP is client.

**Reply by Miq1** [reply]
To get it clear for me: you have your RTU client that wants to get data from the ESP32 #1 TCP server, right? If so, your ESP32 #2 needs to be a RTU bridge that uses a TCP client targeted to ESP32 #1.

You seem to have set up a TCP bridge, that is not able to process requests to it coming in over RTU. You may see a bridge as a certain kind of server, that happens to not serve the requests it gets itself, but will forward these to other servers it knows about. But as a consequence the bridge has to be on the same bus as the initially requesting client is.

**Reply by Miq1** [reply]
Addendum : the line you did not understand is for a TCP bridge only, one that is processing TCP requests. There you may have more than one client requesting, because multiple TCP connections may be open in parallel.

**Reply by Oo87** [reply]
Now make sense !

I used : 

```
// Modbus bridge include
#include "ModbusBridgeWifi.h"
// Create server
ModbusBridgeWiFi MBbridge;
```

instead of : 

```
// Modbus bridge include
#include "ModbusBridgeRTU.h"
// Create server
ModbusBridgeRTU MBbridge(Serial2, 2000,4);
```

I will have to wait tomorow that my inverter switch back on to give a try.
Thanks for the help !, once it works, I will give a complete code example :)

**Reply by Miq1** [reply]
👍🏻

**Comment by Oo87** [comment, marked as answer]
Hello,

Here is the working solution.
The mistake was : using bridgeWifi library instead of bridge RTU.

here is the working code : 

```
// =================================================================================================
// eModbus: Copyright 2020 by Michael Harwerth, Bert Melis and the contributors to ModbusClient
//               MIT license - see license.md for details
// =================================================================================================

#include <Arduino.h>
#include <WiFi.h>
#include "HardwareSerial.h"

// Modbus bridge include for RTU server
#include "ModbusBridgeRTU.h"
// Include the header for the ModbusClient TCP style
#include "ModbusClientTCP.h"

// Give some detailled log
#define LOCAL_LOG_LEVEL LOG_LEVEL_VERBOSE
#include "Logging.h"

WiFiClient theClient;                          // Set up a client
#ifndef MY_SSID
//#define MY_SSID "..."
#define MY_SSID "..."
#endif
#ifndef MY_PASS
#define MY_PASS "..."
#endif

char ssid[] = MY_SSID;                     // SSID and ...
char pass[] = MY_PASS;                     // password for the WiFi network used
uint16_t port = 502;  
                // password for the WiFi network used

// Create a ModbusRTU server instance listening on Serial2 with 2000ms timeout
//ModbusServerRTU MBserver(Serial2, 2000,4);

// Create a ModbusTCP client instance
ModbusClientTCP MB(theClient);

// Create server
ModbusBridgeRTU MBbridge(Serial2, 2000,4);

// Print response from TCP server
void handleData(ModbusMessage response, uint32_t token) 
{
  Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", response.getServerID(), response.getFunctionCode(), token, response.size());
  for (auto& byte : response) {
    Serial.printf("%02X ", byte);
  }
  Serial.println("");
}

// Define an onError handler function to receive error responses
// Arguments are the error code returned and a user-supplied token to identify the causing request
void handleError(Error error, uint32_t token) 
{
  // ModbusError wraps the error code and provides a readable error message for it
  ModbusError me(error);
  Serial.printf("Error response: %02X - %s\n", (int)me, (const char *)me);
}

// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Connect to WiFi
  WiFi.begin(ssid, pass);
  delay(200);
  while (WiFi.status() != WL_CONNECTED) {
    Serial.print(". ");
    delay(1000);
  }
  IPAddress wIP = WiFi.localIP();
  Serial.printf("WIFi IP address: %u.%u.%u.%u\n", wIP[0], wIP[1], wIP[2], wIP[3]);

// Set up ModbusTCP client.
// - provide onData handler function
  MB.onDataHandler(&handleData);
// - provide onError handler function
  MB.onErrorHandler(&handleError);
// Set message timeout to 2000ms and interval between requests to the same host to 200ms
  MB.setTimeout(2000, 200);
// Start ModbusTCP background task
  MB.begin();

// Issue a request => This is useless in bridge mode. it's made to confirm that TCP server is answer at ESP32 startup.
// Set Modbus TCP server address and port number
// (Fill in your data here!)
  MB.setTarget(IPAddress(192, 168, 5, 212), 502);  // Use your adress and port here...

// Create request for
// This is a useless request in order to see if TCP connection to TCP server works.
  Error err = MB.addRequest((uint32_t)millis(), 1, READ_HOLD_REGISTER, 10, 4);
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

// Starting serial on serial 2 (to modbus max485)
  Serial2.begin(9600, SERIAL_8N1, GPIO_NUM_16, GPIO_NUM_17);


// BRIDGE SETUP
// Any request from RTU adress 1 will be send to TCP server with adress 1 too.
  MBbridge.attachServer(1, 1, ANY_FUNCTION_CODE , &MB,IPAddress(192, 168, 5, 212),502);

// Check: print out all combinations served to Serial
  MBbridge.listServer();

// Start the bridge. Port 502, 4 simultaneous clients allowed, 600ms inactivity to disconnect client
  MBbridge.start();


}

void loop() {
  // Restart ESP32 if wifi connection is dead
  delay(5000);
  if (WiFi.status() != WL_CONNECTED) {
    Serial.print("Restarting ESP in 30 seconds");
    delay(30000);
    ESP.restart();
  }
}
```

Thanks for the help ! and great work with eModbus ;)

**Reply by Miq1** [reply]
Glad to read you got it working!

You could clean up a bit and remove the comments from the example you have used as a template. In a few weeks time you may be misled by those if you will have a look again at your code 😉 

(the remarks on the ``onData`` comments are utterly outdated btw - need to clean that up in the library examples... 🙄 )

**Reply by ecertik** [reply]
Hello, The API was changed since this post was send? Because I have a problem with unexpected error messages when I try make wifi bridge based on examples and docs. 
I spend days with this :-( 
Even if I try to just copy and paste this code marked as "working solution" I cant even compile it, because this error message: 
error: 'using ModbusBridgeRTU = class ModbusBridge<ModbusServerRTU>' {aka 'class ModbusBridge<ModbusServerRTU>'} has no member named 'start' MBbridge.start();

When I change it into MBbridge.begin(); I got this error message: 
error: no matching function for call to ModbusBridge<ModbusServerRTU>::begin()'  MBbridge.begin();
candidate: 'void ModbusServerRTU::begin(HardwareSerial&, int)'  void begin(HardwareSerial& serial, int coreID = -1);

When I change it into   MBbridge.begin(Serial0, 1); I got another error message:
ModbusBridgeTemp.h: In instantiation of 'ModbusBridge<SERVERCLASS>::ModbusBridge(HardwareSerial&, uint32_t, int) [with SERVERCLASS = ModbusServerRTU; uint32_t = unsigned int]':
ModbusBridgeTemp.h:89:38: error: no matching function for call to 'ModbusServerRTU::ModbusServerRTU(HardwareSerial&, uint32_t&, int&)'  SERVERCLASS(serial, timeout, rtsPin) { }

etc...

Is actually ANY working example of RTU bridge?

What I need is connect RTU modbus client into ESP board which is via wifi connected into another ESP board where is connected the RTU server.
In short I need exactly modbus wifi bridge. The side with connected RTU server into ESP board looks working from example, but the opposite side is clearly out of my skills. :-(

**Reply by Miq1** [reply]
So you would like to use one ESP to request from a RTU server and serve that over WiFi, right?
If so, you need to incorporate a ``ModbusClientRTU`` into a ``ModbusBridgeWiFi``. The former is started with ``begin()``, the latter with ``start()`` - both with the parameters your local set-up will require. In fact this is what the bridge example in ``examples`` does. It is compiled each time a new library is uploaded to Github and does without errors - else we would not release it.

**Reply by ecertik** [reply]
Hi, thanks for reply.
In the bridge example is side with ModbusClientRTU and ModbusBridgeWiFi. this working fine, I has problems with opposite side, where I needed ModbusClientTCP and ModbusBridgeRTU 
With this opposite side I has problems, because code from first comment in this thread ( https://github.com/eModbus/eModbus/discussions/211#discussioncomment-3372592 ) cant be compiled, it ends with error:
 "'using ModbusBridgeRTU = class ModbusBridge' {aka 'class ModbusBridge'} has no member named 'start' MBbridge.start();"

So I looked into a library, and realized, the class ModbusBridge has not member "start" at all, but instead it have "begin"
That was the easier error. 
I not fully understand how this nice project work, or I found mistake in library, I don't know. Anyway, when I tried register the modbus bridge this way:
ModbusBridgeRTU MBbridge(Serial2, 2000,4);

I got another error about I have unexpected parameter "serial" After I looked into libraries again, I seed this:

ModbusBridgeRTU expecting **three** parameters:   
ModbusBridge(HardwareSerial& serial, uint32_t timeout, int rtsPin = -1) 

after it got **THREE** parameters, it try pass it into ModbusServerRTU:
ModbusBridge<SERVERCLASS>::ModbusBridge(HardwareSerial& serial, uint32_t timeout, int rtsPin) :
  SERVERCLASS(serial, timeout, rtsPin) { }

but ModbusServerRTU expecting **TWO** parameters only:
class ModbusServerRTU : public ModbusServer {
public:
  // Constructors
  explicit ModbusServerRTU(uint32_t timeout, int rtsPin = -1);
  ModbusServerRTU(uint32_t timeout, RTScallback rts); 

I dont know, if I did something by the wrong way, but if I changed in ModbusBridgeTemp.h 
  ModbusBridge(HardwareSerial& serial, uint32_t timeout, int rtsPin = -1);
into this:
  ModbusBridge(uint32_t timeout, int rtsPin = -1);

and this:
ModbusBridge<SERVERCLASS>::ModbusBridge(HardwareSerial& serial, uint32_t timeout, int rtsPin) :
  SERVERCLASS(serial, timeout, rtsPin) { }
into this:
ModbusBridge<SERVERCLASS>::ModbusBridge(uint32_t timeout, int rtsPin) :
  SERVERCLASS(timeout, rtsPin) { }

I successfully compiled the project and it work like a charm now.

**Reply by Miq1** [reply]
You seem to have missed the last few eModbus releases, so it seems. The UART handling has completely changed due to changes in the underlying arduino-esp core.
Your modifications may work for now, but I rather would suggest you downloading the most recent eModbus library and have a look at the changed docs to learn [what has changed with RTU](https://emodbus.github.io/modbusclient-rtu-api).
I will have a look at the ModbusBridge code to see if I may have missed something there 😉

**Comment by emign** [comment]
Hey, sorry to hijack this, but I keep getting confused with the modbus server/client stuff.

Let me explain what I want to do. Maybe you can help me figure out if the solution described in this thread is something for me:

Data source is a controller for an electric wall charger (with modbus 485).  Register table is here: [Modbus RS485.pdf](https://github.com/eModbus/eModbus/files/9331114/Modbus.RS485.pdf). It is supported by this EV charge controller: [evcc](https://github.com/evcc-io/evcc). 
But I need to add it to the software via modbus over LAN/Wifi. It has to read **and** write registers

Is this bridge (RTU client => ESP32 BRIDGE => TCP server) correct for this use case? What confuses me is, that this snippet seems to "just" send to a modbus TCP server and as how I understand it, evcc is not a server but polls the modbus device for information or sends modbus write commands to it.

**Reply by zivillian** [reply]
shameless self plug: https://github.com/zivillian/esp32-modbus-gateway/ is a fully working ModbusTCP Server -> RTU Client Gateway for ESP32 based on eModbus.
One of my use cases is to connect my Heidelberg Wallbox to evcc and it's working great.

**Comment by Miq1** [comment]
@zivillian Nice, I already had a look at it these days 😁
